### PR TITLE
Allow 64-bit --ip-range

### LINKED ISSUE #2296: bitseq: invalid bit range on IPv6
This issue occurred during append IPAMV6Conf to `IPAMConf struct` to automatic allocate IPv6 address as follows:

`invalid bit range [0, 18446744073709551615]`

FILE:
bitseq/sequence.go

### PR DESCRIPTION
**Pull request by robmry**
**- What I did**

Make it possible to use a `--ip-range` that ends on a 64-bit boundary.

Without this change ...

```
# docker network create --ipv6 --subnet fddd::/56 --ip-range fddd::/64 b46
Error response from daemon: failed to allocate gateway (): invalid bit range [0, 18446744073709551615)

# docker network create --ipv6 --subnet fddd::/64 --ip-range fddd::/64 b46
Error response from daemon: failed to allocate gateway (): invalid bit range [0, 18446744073709551615)
```

This is a workaround, although it should solve the issue in practice. I think the remaining issues are unlikely to be a problem (but added `TestIPRangeAt64BitLimit` to illustrate a limitation). A proper fix might be to make the Bitmap understand ranges >=64-bits.

- Fixes https://github.com/moby/libnetwork/issues/2296
  - (Probably! I've not looked at old libnetwork code to check the issue actually is the same.)

**- How I did it**

When `defaultipam.newPoolData` is asked for a range of 64-bits or more, it ends up with an overflowed u64 - so, it just subtracts one to get a nearly-big-enough range (for a 64-bit subnet).

When `defaultipam.getAddress` is called with an `ipr` (sub-pool range), the range it calls `bitmask.SetAnyInRange` with is exclusive of end. So, its end param can't be `MaxUint64`, because that's the max value for the top end of the range and, when checking the range, `SetAnyInRange` fails.

(When `fixed-cidr-v6` behaves more like `fixed-cidr`, it will ask for a 64-bit range if that's what `fixed-cidr-v6` needs.)

Work around that limitation of the Bitmap, by matching the off-by-one when allocating a bit from a range.

The additional check for `ipr == base` avoids the issue in some cases, by ignoring the ipr/sub-pool range if ipr is the same as the pool itself (not really a sub-pool). But, it still fails when `ipr!=base`. For example:

```
docker network create --ipv6 --subnet fddd::/56 --ip-range fddd::/64 b46
```

So, also subtract one from `end` if it's going to hit the max value allowed by the Bitmap.

**- How to verify it**

Without this fix, to see the problem, try:
```
docker network create --ipv6 --subnet fddd::/64 --ip-range fddd::/64 b46
```

New tests, output ...
```
=== RUN   Test64BitIPRange
=== RUN   Test64BitIPRange/64-bit-subnet/no-range/no-gateway
=== RUN   Test64BitIPRange/64-bit-subnet/no-range/with-gateway
=== RUN   Test64BitIPRange/64-bit-subnet/64-bit-range/no-gateway
=== RUN   Test64BitIPRange/64-bit-subnet/64-bit-range/with-gateway
=== RUN   Test64BitIPRange/56-bit-subnet/no-range/no-gateway
=== RUN   Test64BitIPRange/56-bit-subnet/no-range/with-gateway
=== RUN   Test64BitIPRange/56-bit-subnet/64-bit-range/no-gateway
=== RUN   Test64BitIPRange/56-bit-subnet/64-bit-range/with-gateway
--- PASS: Test64BitIPRange (0.82s)
    --- PASS: Test64BitIPRange/64-bit-subnet/no-range/no-gateway (0.09s)
    --- PASS: Test64BitIPRange/64-bit-subnet/no-range/with-gateway (0.10s)
    --- PASS: Test64BitIPRange/64-bit-subnet/64-bit-range/no-gateway (0.10s)
    --- PASS: Test64BitIPRange/64-bit-subnet/64-bit-range/with-gateway (0.11s)
    --- PASS: Test64BitIPRange/56-bit-subnet/no-range/no-gateway (0.10s)
    --- PASS: Test64BitIPRange/56-bit-subnet/no-range/with-gateway (0.11s)
    --- PASS: Test64BitIPRange/56-bit-subnet/64-bit-range/no-gateway (0.11s)
    --- PASS: Test64BitIPRange/56-bit-subnet/64-bit-range/with-gateway (0.09s)
=== RUN   TestIPRangeAt64BitLimit
=== RUN   TestIPRangeAt64BitLimit/ipRange_before_end_of_64-bit_subnet
=== RUN   TestIPRangeAt64BitLimit/ipRange_at_end_of_64-bit_subnet
    bridge_test.go:196: XFAIL: Container startup failed with error: Error response from daemon: no available IPv6 addresses on this network's address pools: testnet (6e561e06c66648b1baaa5cb20fd474ee79c98907466ef77e0853537741ebb20c)
=== RUN   TestIPRangeAt64BitLimit/ipRange_at_64-bit_boundary_inside_56-bit_subnet
    bridge_test.go:196: XFAIL: Container startup failed with error: Error response from daemon: no available IPv6 addresses on this network's address pools: testnet (4ee6fa98e8d1feeb2a92e1ffa7f1e422e42174dcce33546fd29d03b60061aad8)
--- PASS: TestIPRangeAt64BitLimit (0.53s)
    --- PASS: TestIPRangeAt64BitLimit/ipRange_before_end_of_64-bit_subnet (0.29s)
    --- SKIP: TestIPRangeAt64BitLimit/ipRange_at_end_of_64-bit_subnet (0.12s)
    --- SKIP: TestIPRangeAt64BitLimit/ipRange_at_64-bit_boundary_inside_56-bit_subnet (0.11s)
```

**- Description for the changelog**
```markdown changelog
n/a - workaround, replaced by https://github.com/moby/moby/pull/49223
```

### CODE REVIEW

File: integration/network/bridge_test.go [hunk 1]
```diff
@@ -97,3 +99,104 @@ func TestCreateWithIPv6WithoutEnableIPv6Flag(t *testing.T) {
...
 }
+
+// Check that it's possible to create IPv6 networks with a 64-bit ip-range,
+// in 64-bit and bigger subnets, with and without a gateway.
+func Test64BitIPRange(t *testing.T) {
+	skip.If(t, testEnv.DaemonInfo.OSType == "windows", "no bridge or IPv6 on Windows")
+	ctx := setupTest(t)
+	c := testEnv.APIClient()
+
+	type kv struct{ k, v string }
```
Comment (akerouanton) [inline integration/network/bridge_test.go:110]: Why not use a `map[string]string` instead?

Comment (robmry) [reply integration/network/bridge_test.go:110]: Just to make the order of the tests predictable.

**Review by akerouanton — APPROVED** [review]

File: integration/network/bridge_test.go [hunk 2]
```diff
@@ -97,3 +99,104 @@ func TestCreateWithIPv6WithoutEnableIPv6Flag(t *testing.T) {
...
+		{"64-bit-range", "fd2e:b68c:ce26::/64"},
+	}
+	gateways := []kv{
+		{"no-gateway", ""},
+		{"with-gateway", "fd2e:b68c:ce26::1"},
+	}
+
+	for _, sn := range subnets {
+		for _, ipr := range ipRanges {
+			for _, gw := range gateways {
```
[the remark this answers is not in the document]
Comment (robmry) [reply integration/network/bridge_test.go:126]: I don't think so - the vars aren't used inside the `Run()`.

File: integration/network/bridge_test.go [hunk 3]
```diff
@@ -97,3 +99,104 @@ func TestCreateWithIPv6WithoutEnableIPv6Flag(t *testing.T) {
...
+				ipamSetter := network.WithIPAMRange(sn.v, ipr.v, gw.v)
+				t.Run(sn.k+"/"+ipr.k+"/"+gw.k, func(t *testing.T) {
+					ctx := testutil.StartSpan(ctx, t)
+					const netName = "test64br"
+					network.CreateNoError(ctx, t, c, netName, network.WithIPv6(), ipamSetter)
+					defer network.RemoveNoError(ctx, t, c, netName)
```
[the remark this answers is not in the document]
Comment (robmry) [reply integration/network/bridge_test.go:132]: Mostly just habit, but I did think about starting/inspecting a container ... maybe we'll want to do that at some point.

File: libnetwork/ipams/defaultipam/allocator.go [hunk 4]
```diff
@@ -215,6 +216,7 @@ func newPoolData(pool netip.Prefix) *PoolData {
 
 	// Allow /64 subnet
 	if pool.Addr().Is6() && numAddresses == 0 {
+		// FIXME(robmry) - see the comment in getAddress
 		numAddresses--
 	}
```
Comment (thaJeztah) [inline libnetwork/ipams/defaultipam/allocator.go:221]: OHMAN yeah, this is confusing `numAddresses == 0` as the condition .. so let's decrement :confused_travolta:

**Review by thaJeztah — APPROVED** [review]
LGTM

**Pull request merged**